src/dragonball/src/device_manager/mod.rs

```diff
@@ -0,0 +1,600 @@
+            irq_manager: Arc::new(KvmIrqManager::new(vm_fd.clone())),
+            res_manager,
+            vm_fd,
+            logger: logger.new(slog::o!()),
+            con_manager: ConsoleManager::new(epoll_manager, logger),
+            legacy_manager: None,
+            #[cfg(feature = "virtio-vsock")]
+            vsock_manager: VsockDeviceMgr::default(),
+        }
+    }
+
+    /// Create the underline interrupt manager for the device manager.
+    pub fn create_interrupt_manager(&mut self) -> Result<()> {
+        self.irq_manager
+            .initialize()
+            .map_err(DeviceMgrError::CreateDevice)
+    }
+
+    /// Get the underlying logger.
+    pub fn logger(&self) -> &slog::Logger {
+        &self.logger
+    }
+
+    /// Create legacy devices associted virtual machine
+    pub fn create_legacy_devices(
+        &mut self,
+        ctx: &mut DeviceOpContext,
+    ) -> std::result::Result<(), StartMicrovmError> {
+        #[cfg(target_arch = "x86_64")]
+        {
+            let mut tx = ctx.io_context.begin_tx();
+            let legacy_manager =
+                LegacyDeviceManager::create_manager(&mut tx.io_manager, Some(self.vm_fd.clone()));
+
+            match legacy_manager {
+                Ok(v) => {
+                    self.legacy_manager = Some(v);
+                    ctx.io_context.commit_tx(tx);
+                }
+                Err(e) => {
+                    ctx.io_context.cancel_tx(tx);
+                    return Err(StartMicrovmError::LegacyDevice(e));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Init legacy devices with logger stream in associted virtual machine
+    pub fn init_legacy_devices(
+        &mut self,
+        dmesg_fifo: Option<Box<dyn io::Write + Send>>,
+        com1_sock_path: Option<String>,
+        _ctx: &mut DeviceOpContext,
+    ) -> std::result::Result<(), StartMicrovmError> {
+        // Connect serial ports to the console and dmesg_fifo.
+        self.set_guest_kernel_log_stream(dmesg_fifo)
+            .map_err(|_| StartMicrovmError::EventFd)?;
+
+        slog::info!(self.logger, "init console path: {:?}", com1_sock_path);
+        if let Some(path) = com1_sock_path {
+            if let Some(legacy_manager) = self.legacy_manager.as_ref() {
+                let com1 = legacy_manager.get_com1_serial();
+                self.con_manager
+                    .create_socket_console(com1, path)
+                    .map_err(StartMicrovmError::DeviceManager)?;
+            }
+        } else if let Some(legacy_manager) = self.legacy_manager.as_ref() {
+            let com1 = legacy_manager.get_com1_serial();
+            self.con_manager
+                .create_stdio_console(com1)
+                .map_err(StartMicrovmError::DeviceManager)?;
+        }
+
+        Ok(())
+    }
+
+    /// Set the stream for guest kernel log.
+    ///
+    /// Note: com2 is used for guest kernel logging.
+    /// TODO: check whether it works with aarch64.
```

Comment:
Maybe make this a GitHub issue URL as it's easy to forget TODO's and FIXME's in the code :smile:

@jongwu - fyi ^

Reply:
I'm glad to see that Dragonball can support aarch64 and happy to test it on arm side, but for now I don't know how to start. Is there any guide to get me there? @studychao , thanks.

Reply:
Thanks, @jongwu :) 
Dragonball is still under development so it's hard to give a version for you to test it right now. We plan to submit all the code before the end of June, at that time you could have a try at it.
I'll let you know when it's ready for you to test and try~

Reply:
Thanks, @jongwu .

Arm Dragonball is part of rund, and rund is still under development, so it cannot be tested separately. We are glad to invite you to join us to improve its capabilities. For now, by adding unit tests, you may test some specific mechanisms on arm, like fdt, gic, its, vcpu and relevent procedures. We are open for any questions and feedbacks on arm Dragonball.

Reply:
Sounds good. I can do some tests manually and it's better to add jenkins ci job for the runtime-rs in the future.

Reply:
Thanks, @jodh-intel :)

The GitHub issue URL for this TODO is #4409 .
We will complete this work in soon.